### ISSUE TITLE
wildcard_enum_match_arm is not very helpful in combination with non_exhaustive

### ISSUE THREAD

**Issue by pacak** [rust-lang/rust-clippy#8540]
### Summary

Currently `wildcard_enum_match_arm` will suggest to fill in all the existing variants in addition to the wildcard one. Motivation for the lint is "New enum variants added by library updates can be missed", but sad reality of `non_exhaustive` means new enum variants will be always missed regardless of the code I'm writing

### Lint Name

 wildcard_enum_match_arm 

### Reproducer

I tried this code:

```rust
        let arr = match &ty {
            Type::Array(arr) => Some(arr),
            _ => None,
        };
```

I saw this happen:

```
    |
481 |             _ => None,
    |             ^ help: try this: `Type::BareFn(_) | Type::Group(_) | Type::ImplTrait(_) | Type::Infer(_) | Type::Macro(_) | Type::Never(_) | Type::Paren(_) | Type::Path(_) | Type::Ptr(_) | Type::Reference(_) | Type::Slice(_) | Type::TraitObject(_) | Type::Tuple(_) | Type::Verbatim(_) | _`
    |

```

I expected to see this happen: no suggestion at all. Proposed change fails to achieve the goal and only adds noise.

### Version

```text
rustc 1.58.0 (02072b482 2022-01-11)
binary: rustc
commit-hash: 02072b482a8b5357f7fb5e5637444ae30e423c40
commit-date: 2022-01-11
host: x86_64-unknown-linux-gnu
release: 1.58.0
LLVM version: 13.0.0
```

**Comment by oskgo** [comment]
For non-exhaustive enums there are two cases, you get a variant you know the existence of, or you get a variant you don't know the existence of. The ones you don't know of have to be handled with the wildcard pattern.

The lint is telling you that you know about more enum variants than you are currently explicitly handling.

Sometimes you'll want to handle new variants explicitly as soon as they are introduced, especially when your only option for the wildcard is to return an error or panic (unimplemented!).

I think the occasional noise is worth it unless your situation (mapping unknown and most known variants to the same value) is a lot more common than the one I outlined.

**Comment by pacak** [comment]
> Sometimes you'll want to handle new variants explicitly as soon as they are introduced, especially when your only option for the wildcard is to return an error or panic (unimplemented!).

Right, I generally want to handle unknown variants, but for handling them I would prefer to have something like `#[ignore_non_exhaustive]` on match and turning it into a compile error any time something changes. Relying on a lint seems odd.

**Comment by oskgo** [comment]
You can try to get it added to the language. I think there has been some ideas along those lines in the past, though there were some objections (adding such an attribute would break semver and turn non-breaking changes into breaking ones).

Regardless, I think the lint should be kept until such a feature is in place, unless there's too many false-positives or something else is wrong about the lint itself.

**Comment by pacak** [comment]
Fair enough, I'll see what I can do.
